### PR TITLE
Enhancements for Login Password input.

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Enhancements to better handle the password string input and make it more user-friendly by providing feedback, indicating the current string length & allowing users to toggle the "visibility" of the real password vs showing an asterisk for each character.

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

When you get a chance, please review & validate the following functionality:

- Provide feedback as each character is typed. By default, an asterisk is displayed for each char.
- The Backspace key works as usual to delete previous characters.
- The TAB key works as a toggle to make the real password string visible or show asterisks.
- Only 7-bit ASCII printable chars are considered valid.
- The current length of the password string is also shown/updated as you type or delete chars.
- The maximum length allowed is 64 chars (which is double the max length allowed by ASUS).
- You can use Copy&Paste to enter the password and the entry will show feedback as well.
- The current password entry stored in the configuration file will be shown as default.

**Comment by ExtremeFiretop** [comment]
> Enhancements to better handle the password string input and make it more user-friendly by providing feedback, indicating the current string length & allowing users to toggle the "visibility" of the real password vs showing an asterisk for each character.

9AM my time and I just woke up. Did you sleep at all last night? hahahaha!

Let me review this after my quick meeting this morning and merge, so far quickly looks great!

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> When you get a chance, please review & validate the following functionality:
> 
> * Provide feedback as each character is typed. By default, an asterisk is displayed for each char.
> * The Backspace key works as usual to delete previous characters.
> * The TAB key works as a toggle to make the real password string visible or show asterisks.
> * Only 7-bit ASCII printable chars are considered valid.
> * The current length of the password string is also shown/updated as you type or delete chars.
> * The maximum length allowed is 64 chars (which is double the max length allowed by ASUS).
> * You can use Copy&Paste to enter the password and the entry will show feedback as well.
> * The current password entry stored in the configuration file will be shown as default.

Holy fancy! I love it. I'll test all these features individually

**Comment by Martinski4GitHub** [comment]
> 
> 9AM my time and I just woke up. Did you sleep at all last night? hahahaha!
> 
> Let me review this after my quick meeting this morning and merge, so far quickly looks great!
>
I went to sleep a little past midnight. I had the function completed, but I had not yet fully tested it so I woke up about half an hour ago and did testing & validation before letting you take over and do your usual review & testing. I really like the way it turned out (if I do say so myself). It makes inputting the password much more user-friendly with the feedback & how you can toggle the visibility of the real password. I have a 28-char passphrase and this makes it very easy to just copy/paste it and then verify that I got it all correct by pressing the TAB char before hitting the Enter key.

**Comment by ExtremeFiretop** [comment]
> > 9AM my time and I just woke up. Did you sleep at all last night? hahahaha!
> > Let me review this after my quick meeting this morning and merge, so far quickly looks great!
> 
> I went to sleep a little past midnight. I had the function completed, but I had not yet fully tested it so I woke up about half an hour ago and did testing & validation before letting you take over and do your usual review & testing. I really like the way it turned out (if I do say so myself). It makes inputting the password much more user-friendly with the feedback & how you can toggle the visibility of the real password. I have a 28-char passphrase and this makes it very easy to just copy/paste it and then verify that I got it all correct by pressing the TAB char before hitting the Enter key.

Enjoy your day buddy, your always great! Next time you hear from me I'll have reviewed this for our next release! Keep your head above water these next few months!

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

I have another engineering meeting earlier today in about 1 hour (with one of our European offices), and then another one after lunch, plus other stuff I have to do before Monday, so I'll be busy.

I'll try to do a quick check on messages whenever I can, but I'll be mostly offline for the rest of the day.

File: MerlinAU.sh [hunk 1]
```diff
@@ -1905,7 +1905,7 @@ _TestLoginCredentials_()
         printf "\n${GRNct}Login test passed.${NOct}"
         "$isInteractive" && printf "\nRestarting web server... Please wait.\n"
         /sbin/service restart_httpd >/dev/null 2>&1 &
-        sleep 5
+        sleep 1
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1908]: Approved

File: MerlinAU.sh [hunk 2]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
     fi
 }
 
-##------------------------------------------##
-## Modified by ExtremeFiretop [2024-Feb-29] ##
-##------------------------------------------##
+##-------------------------------------##
+## Added by Martinski W. [2024-Feb-29] ##
+##-------------------------------------##
+_GetPasswordInput_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1923]: Approved

File: MerlinAU.sh [hunk 3]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
...
+
+    while [ "$retry" = "yes" ]
+    do
+        echo "=== Login Credentials ==="
+        _GetPasswordInput_ "Enter password for user ${GRNct}${userName}${NOct}"
+        if [ -z "$pswdString" ]
         then
-            echo "Password cannot be empty. Credentials were not saved."
-            _WaitForEnterKey_ "$mainMenuReturnPromptStr"
+            printf "\nPassword string is ${REDct}NOT${NOct} valid. Credentials were not saved.\n"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2060]: Approved

File: MerlinAU.sh [hunk 4]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
...
-        username="$(nvram get http_username)"
+    # Get the Username from NVRAM #
+    userName="$(nvram get http_username)"
 
-        # Prompt the user only for a password [-s flag hides the password input]
-        printf "Enter password for user ${GRNct}${username}${NOct}: "
-        read -rs password
-        echo
-        if [ -z "$password" ]
+    loginCredsENC="$(Get_Custom_Setting credentials_base64)"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2045]: Approved

File: MerlinAU.sh [hunk 5]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
...
+   local retCode  charNum  prevChar  pswdLength  showPSWD
+
+   if [ $# -eq 0 ] || [ -z "$1" ]
+   then
+       printf "${REDct}**ERROR**${NOct}: NO prompt string was provided.\n"
+       return 1
+   fi
+   PSWDprompt="$1"
+
+   _showPSWDPrompt_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1936]: Approved

File: MerlinAU.sh [hunk 6]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
...
+
+      ## TAB keypress is a toggle to make password string "visible" ##
+      if [ "$charNum" -eq 9 ]
+      then
+          showPSWD="$((! showPSWD))"
+          _showPSWDPrompt_
+          continue
+      fi
+
+      ## Ignore Escape Sequences ##
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1990]: Aproved

File: MerlinAU.sh [hunk 7]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
...
+              PSWDstring=""
+              printf "\n${REDct}**ERROR**${NOct}: Password length is greater than allowed maximum length "
+              printf "[MAX=${GRNct}${PSWDstrLenMAX}${NOct}].\n"
+              retCode=1
+          fi
+          break
+      fi
+      charNum="$(printf "%d" "'$theChar")"
+
+      ## TAB keypress is a toggle to make password string "visible" ##
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1982]: Approved

File: MerlinAU.sh [hunk 8]
```diff
@@ -1917,43 +1917,164 @@ _TestLoginCredentials_()
...
+          then
+              PSWDtmpStr="${PSWDtmpStr%?}"
+              PSWDstring="${PSWDstring%?}"
+              pswdLength="$((pswdLength - 1))"
+              _showPSWDPrompt_
+              continue
+          fi
+      fi
+
+      ## ONLY 7-bit ASCII printable characters are VALID ##
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2017]: Approved

**Comment by ExtremeFiretop** [comment]
Sorry for the delay, I started reviewing and then got called into another meeting (Department is moving to WuFB)
It's the talk of the team right now lol.

So for the review, everything went well. Same for testing!
Here are the results below:

1. Provide feedback as each character is typed. By default, an asterisk is displayed for each char.
**= Tested and confirmed functional**

2. The Backspace key works as usual to delete previous characters.
**= Tested and confirmed functional**

3. The TAB key works as a toggle to make the real password string visible or show asterisks.
**= Tested and confirmed functional. **
However, I noticed if the user holds it for some reason as a "hold to view" instead of a "tab to view" then it starts flicking at a million miles a minute lol!

4. Only 7-bit ASCII printable chars are considered valid.
**= Tested and confirmed functional**

6. The current length of the password string is also shown/updated as you type or delete chars.
**= Tested and confirmed functional**

7. The maximum length allowed is 64 chars (which is double the max length allowed by ASUS).
**= Tested and confirmed functional**

8. You can use Copy&Paste to enter the password and the entry will show feedback as well.
**= Tested and confirmed functional**

10. The current password entry stored in the configuration file will be shown as default.
**= Tested and confirmed functional**

**Comment by ExtremeFiretop** [comment]
Created PR: 154 for the TAB debounce issue:
https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/154

File: MerlinAU.sh (hunk 7 above)
Comment (ExtremeFiretop) [reply MerlinAU.sh:1982]: Submitted PR: 154 to address a possible user mis-using this by holding TAB.

**Comment by Martinski4GitHub** [comment]
> 
>     3. The TAB key works as a toggle to make the real password string visible or show asterisks.
>        **= Tested and confirmed functional**
>        However, I noticed if the user holds it for some reason as a "hold to view" instead of a "tab to view" then it starts flicking at a million miles a minute lol!
> 
Man, you like to abuse your system, don't you? LOL :>).
But yeah, I see what you mean. Good find!!

**Comment by ExtremeFiretop** [comment]
> > ```
> > 3. The TAB key works as a toggle to make the real password string visible or show asterisks.
> >    **= Tested and confirmed functional**
> >    However, I noticed if the user holds it for some reason as a "hold to view" instead of a "tab to view" then it starts flicking at a million miles a minute lol!
> > ```
> 
> Man, you like to abuse your system, don't you? LOL :>). But yeah, I see what you mean. Good find!!

I just like pushing things a bit. Remember I worked IT Support for YEARS before I moved into Engineering only in the last few years. I've seen all the dumb things users can do and complain about!

**Comment by ExtremeFiretop** [comment]
I should mention I got a test router specifically for testing our scripts now that doesn't impact my production internet now that it's all functioning normally haha! So I'm all about abusing it now 😜

**Comment by Martinski4GitHub** [comment]
> 
> I've seen all the dumb things users can do and complain about!
>
I hear you!! I've had conversations with people from the customer support group as well and the stories they tell us can be both hilarious & frustrating at the same time. Classic PEBKAC stories!! LOL.

**Comment by Martinski4GitHub** [comment]
> I should mention I got a test router specifically for testing our scripts so it doesn't impact my production internet now that it's all functioning normally haha! So I'm all about abusing it now 😜\
>
Good call. I tend to be gentle with my router, hoping to get another year before I need to replace it (so far it's been solid and good to us).

**Pull request merged**